installed script lacks "-E" in their shebang

I installed `pyflakes` through pipx, but using the tool in another context fails, because that other context set a PYTHONPATH that interfer with `pyflakes` import. 

<!-- Please be as detailed as possible! -->

**How to reproduce**

`pyflakes` installed with `Python 3.12`, somehow a script added an element to `PYTHONPATH` that contains a re module compile for `Python 3.11`.  pyflakes wrongly pick up the PYTHON PATH instead of sticking within its venv.

```
   $ pyflakes --version
+  Traceback (most recent call last):
+    File "HOME/.local/bin//pyflakes", line 3, in <module>
+      import re
+    File "OTHER/re/__init__.py", line 125, in <module>
+      from . import _compiler, _parser
+    File "OTHER/_compiler.py", line 18, in <module>
+      assert _sre.MAGIC == MAGIC, "SRE module mismatch"
+             ^^^^^^^^^^^^^^^^^^^
+  AssertionError: SRE module mismatch
+  [1]
```

**Expected behavior**

I would expect pipx to install standalone, isolated environment that are not affected by external factor. This expectation fails here.


**Possible fixes ?**

Install tools with `-E` in their shebang for appropriate isolation. I fixed my install that way.

--- a/changelog.d/1584.bugfix.md
+++ b/changelog.d/1584.bugfix.md
@@ -1,0 +1,1 @@
+Install app scripts with shebangs that ignore `PYTHONPATH`.

--- a/src/pipx/commands/common.py
+++ b/src/pipx/commands/common.py
@@ -60,6 +60,8 @@ def expose_resources_globally(
     suffix: str = "",
 ) -> None:
     for path in paths:
+        if resource_type == "app":
+            _add_ignore_environment_to_python_shebang(path)
         src = path.resolve()
         if resource_type == "man":
             dest_dir = local_resource_dir / src.parent.name
@@ -100,6 +102,28 @@ def can_symlink(local_resource_dir: Path) -> bool:
                 _can_symlink_cache[local_resource_dir] = False
 
     return _can_symlink_cache[local_resource_dir]
+
+
+def _add_ignore_environment_to_python_shebang(path: Path) -> None:
+    if WINDOWS or not path.is_file():
+        return
+
+    try:
+        data = path.read_bytes()
+    except OSError:
+        return
+
+    first_line, separator, rest = data.partition(b"\n")
+    if not first_line.startswith(b"#!"):
+        return
+
+    interpreter = first_line[2:]
+    if interpreter.endswith(b" -E"):
+        return
+    if b"python" not in interpreter.lower() or b" " in interpreter or b"\t" in interpreter:
+        return
+
+    path.write_bytes(first_line + b" -E" + separator + rest)
 
 
 def _copy_package_resource(dest_dir: Path, path: Path, suffix: str = "") -> None:

--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -1,5 +1,9 @@
+import os
+import subprocess
+import sys
+
 from helpers import skip_if_windows
-from pipx.commands.common import get_exposed_paths_for_package
+from pipx.commands.common import expose_resources_globally, get_exposed_paths_for_package
 
 
 @skip_if_windows
@@ -14,3 +18,35 @@ def test_get_exposed_paths_ignores_recursive_symlink(tmp_path):
     exposed = get_exposed_paths_for_package(venv_resource_path, local_resource_dir)
 
     assert loop not in exposed
+
+
+@skip_if_windows
+def test_expose_app_scripts_ignores_pythonpath(tmp_path):
+    venv_resource_path = tmp_path / "venv_bin"
+    venv_resource_path.mkdir()
+    local_resource_dir = tmp_path / "bin"
+    shadow_path = tmp_path / "shadow"
+    shadow_path.mkdir()
+
+    app_path = venv_resource_path / "demo"
+    app_path.write_text(
+        f"#!{sys.executable}\n"
+        "import sys\n"
+        f"if {str(shadow_path)!r} in sys.path:\n"
+        "    raise SystemExit('PYTHONPATH leaked into app script')\n"
+        "print('ok')\n"
+    )
+    app_path.chmod(0o755)
+
+    expose_resources_globally("app", local_resource_dir, [app_path], force=False)
+
+    assert app_path.read_text().splitlines()[0] == f"#!{sys.executable} -E"
+
+    result = subprocess.run(
+        [app_path],
+        check=True,
+        capture_output=True,
+        env={**os.environ, "PYTHONPATH": str(shadow_path)},
+        text=True,
+    )
+    assert result.stdout == "ok\n"

